### DISCUSSION TITLE
posts/how-reality-works/

### DISCUSSION THREAD

**Discussion by giscus** [ShadowObj/blog#4] [Announcements]
# posts/how-reality-works/

超越常规TLS的安全性? 无需买域名、配置TLS? \

https://objshadow.pages.dev/posts/how-reality-works/

**Comment by ObjShadow** [comment]
由于本博客使用的是hugo框架，昨天(2024-08-07)文章发布时hugo将代码块内的代码也进行了统计，故而统计得到的字数来到了夸张的约10500词，今天已进行了修复，望读者见谅。

**Comment by seeker252** [comment]
Hello.
Today morning I thought, that I found some serious Reality volnurability. I hope to listen from you that I mistaken.
Look. Reality works as TLS 1.3 server and use SessionID for authenticate valid client connections. All requests from other clients transparently redirected to remote server by TCP. In TLS 1.3 SessionID must be random ir zero. Any TLS 1.3 server don't validate it. Ok. If third party (censor) change Reality client SessionID (with encrypted authentication data) in ClientHello with any 32 bit data for all TLS 1.3 Hellos to foreign IPs or IPs from grey lists, I think all legal (non-Reality) TLS 1.3 connections not be violated. We even can connect with our Reality VPS, but Reality server due to Session ID changed by third party will think about us as not valid Reality client and send our connection to remote server.
So censor no need to find Reality servers specially. He may mechanically change Session ID of TLS 1.3 ClientHello to any randon 32-bit value for some connections to some IPs and not violate any TLS 1.3 non-Reality connections. So HTTPs will work fine, but Reality don't.
Or I mistaken?

**Reply by ObjShadow** [reply]
Hello Guest! Thanks for your visit and comment.

You argue that since the REALITY protocol uses the SessionID field in the TLS 1.3 ClientHello to distinguish legitimate REALITY clients from others, a man-in-the-middle (MITM) censor could block all REALITY connections by modifying this field without affecting normal TLS traffic.  

However, this is **impossible**. In TLS 1.3, the key derivation process at every stage of the handshake incorporates the cryptographic hash of all preceding TLS records (including ClientHello, ServerHello, etc.). Since the SessionID field is part of the ClientHello message, any tampering with it would alter the hash value of the ClientHello.

This would result in key derivation failure for the tampered TLS 1.3 connection, ultimately causing the **handshake to abort**.
